### PR TITLE
feat(core) CDB-1952: Warn if no pubsub messages

### PR DESCRIPTION
**Pull request by dbcfd**
Log a warning if pubsub messages are not timely. Will warn if no messages come from ipfs, or only external messages come from IPFS.

## Description

Include a summary of the change or link to the issue it addresses.

Include relevant motivation, context, brief description and impact of the change(s). List follow-up tasks here.

## How Has This Been Tested?

Describe the tests that you ran to verify your changes. Provide instructions for reproduction.

- [ ] Test A (e.g. Test A - New test that ... ran in local, docker, and dev unstable.)
- [ ] Test B

## PR checklist

Before submitting this PR, please make sure:

- [ ] I have tagged the relevant reviewers and interested parties
- [ ] I have updated the READMEs of affected packages
- [ ] I have made corresponding changes to the documentation

## References:

Please list relevant documentation (e.g. tech specs, articles, related work etc.) relevant to this change, and note if the documentation has been updated.

### CODE REVIEW

File: packages/core/src/pubsub/__tests__/pubsub.test.ts [hunk 1]
```diff
@@ -21,7 +23,7 @@ const PEER_ID = 'PEER_ID'
 const LENGTH = 2
 const MESSAGES = Array.from({ length: LENGTH }).map((_, index) => {
   return {
-    typ: MsgType.QUERY as MsgType.QUERY,
+    typ: <MsgType.QUERY>MsgType.QUERY,
```
Comment (ukstv) [inline packages/core/src/pubsub/__tests__/pubsub.test.ts]: This seems like an old way of casting in TS. I might be wrong here.

File: packages/core/src/pubsub/pubsub.ts [hunk 2]
```diff
@@ -53,15 +53,20 @@ export class Pubsub extends Observable<PubsubMessage> {
...
   ) {
     super((subscriber) => {
-      const incoming$ = new IncomingChannel(ipfs, topic, resubscribeEvery, pubsubLogger, logger, tasks)
+      const incoming$ = new IncomingChannel(ipfs, topic, resubscribeEvery, lateMessageAfterSeconds, pubsubLogger, logger, tasks)
 
       incoming$
-        .pipe(filterExternal(this.peerId$), ipfsToPubsub(this.peerId$, pubsubLogger, topic))
+        .pipe(
+          filterExternal(this.peerId$),
+          checkSlowObservable(lateMessageAfterSeconds * 1000, logger, `IPFS did not provide any internal messages, please check your IPFS configuration.`),
```
Comment (ukstv) [inline packages/core/src/pubsub/pubsub.ts]: It is applied twice IMO. The first time it is applied in IncomingChannel itself.

File: packages/core/src/pubsub/incoming-channel.ts [hunk 3]
```diff
@@ -103,3 +106,32 @@ export function filterExternal(
     map((data) => data.entry)
   )
 }
+
+export function checkSlowObservable(
+  delay: number,
+  logger: DiagnosticsLogger,
+  description: string,
+): MonoTypeOperatorFunction<IPFSPubsubMessage> {
+  let outstanding = asyncScheduler.schedule(
```
Comment (ukstv) [inline packages/core/src/pubsub/incoming-channel.ts]: Overall, it seems like a right approach. I mean, unless someone finds an operator like that in rxjs upstream.

File: packages/core/src/pubsub/incoming-channel.ts [hunk 4]
```diff
@@ -103,3 +106,32 @@ export function filterExternal(
...
+    () => {
+      logger.warn(`Message was not timely. ${description}`)
+    },
+    delay
+  )
+  return pipe(
+    tap(() => {
+      if(outstanding != null) {
```
Comment (ukstv) [inline packages/core/src/pubsub/incoming-channel.ts]: outstanding is never null, unless I miss something, so this is a noop semantically. Worth removing the if condition.

File: packages/core/src/pubsub/incoming-channel.ts [hunk 5]
```diff
@@ -103,3 +106,32 @@ export function filterExternal(
...
+        outstanding.unsubscribe()
+      }
+      const sub = asyncScheduler.schedule(
+        () => {
+          if(outstanding == sub) {
```
Comment (ukstv) [inline packages/core/src/pubsub/incoming-channel.ts]: I think you could drop this if condition.

If I understand the code and the intent behind it correctly, here we move a time window* for the warning every time a message appears.

* - at the end of the window, we log a warning.

This means, again unless I miss something, that if we have a steady stream of messages, we at the end of the window started by a message have outstanding != sub. We have multiple overlapping windows.

My best guess, is that the condition could be dropped if we tell "sub" to unsubscribe (= stop waiting for the timeout) before garbage collecting it.

Comment (ukstv) [reply packages/core/src/pubsub/incoming-channel.ts]: Before doing outstanding = sub, do something like sub.unsubscribe().

Comment (ukstv) [reply packages/core/src/pubsub/incoming-channel.ts]: Oh, you do unsubscribe before. So, I do not see the point of having the condition here.

File: packages/core/src/pubsub/incoming-channel.ts (hunk 3 above)
Comment (ukstv) [reply packages/core/src/pubsub/incoming-channel.ts]: Out of curiosity, why have you chosen asyncScheduler? Why not good old setTimeout/clearTimeout?

File: packages/core/src/pubsub/__tests__/pubsub.test.ts [hunk 6]
```diff
@@ -78,3 +80,145 @@ test('publish', async () => {
...
+  const ipfs = {
+    pubsub: {
+      subscribe: async (_, handler) => {
+        feed.subscribe(handler)
+      },
+      unsubscribe: jest.fn(),
+      ls: jest.fn(() => []),
+      publish: jest.fn(),
+    },
+    id: jest.fn(async () => ({ id: PEER_ID })),
```
Comment (ukstv) [inline packages/core/src/pubsub/__tests__/pubsub.test.ts:153]: Worth deduplicating this in a function?

File: packages/core/src/pubsub/__tests__/pubsub.test.ts (hunk 1 above)
Comment (dbcfd) [reply packages/core/src/pubsub/__tests__/pubsub.test.ts]: Thanks for seeing that. Was an issue with my build environment, and failed to revert that change.

File: packages/core/src/pubsub/pubsub.ts (hunk 2 above)
Comment (dbcfd) [reply packages/core/src/pubsub/pubsub.ts]: So incoming channel is checking for just a general misconfiguration with no messages from ipfs, while this checks for a misconfiguration and only receiving external messages. Bad part is, you would get two log messages.

**Review by ukstv — APPROVED** [review]

**Comment by dbcfd** [comment]
Closing for https://github.com/ceramicnetwork/js-ceramic/pull/2671

**Pull request closed without merging**